### DISCUSSION TITLE
Linker error in make, installing iraf-2.17 on macintel

### DISCUSSION THREAD

**Discussion by janerigby** [iraf-community/iraf#206] [Q&A] [answered]
Here I am at the car dealership, trying to install Iraf so I can flux calibrate some Magellan spectra.  I'm getting a weird set of linker errors, copied below.  Advice appreciated; I've never seen this before.  Thanks!  -Jane

```
----------------------- GENERIC ------------------------
+ cd generic
+ sh -x mkpkg.sh
+ set -e
+ flex -o lexyy.c tok.l
+ gcc -c -g -Wall -O2 -I/Users/jrrigby1/Applications/iraf-2.17/include generic.c yywrap.c lexyy.c
+ gcc generic.o lexyy.o yywrap.o /Users/jrrigby1/Applications/iraf-2.17/unix/hlib/libboot.a /Users/jrrigby1/Applications/iraf-2.17/lib/libsys.a /Users/jrrigby1/Applications/iraf-2.17/lib/libvops.a /Users/jrrigby1/Applications/iraf-2.17/unix/hlib/libos.a /Users/jrrigby1/Applications/iraf-2.17/unix/bin.macintel/libf2c.a -lm -o generic.e
ld: file too small (length=8) file '/Users/jrrigby1/Applications/iraf-2.17/lib/libsys.a' for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [sysgen] Error 1
```
So let's see what's in that libsys.a file:
```
% more libsys.a 
!<arch>
```

**Comment by olebole** [comment]
Please check that `lib/libsys.a` is a symbolic link and points to `../bin/libsys.a`. This is a common problem when you used the zip file instead of the tarball for the sources -- you really need [the **tarball**](https://github.com/iraf-community/iraf/archive/refs/tags/v2.17.tar.gz) (or at least unpack it with the command line zip, not with the GUI).

Otherwise, can you add the full log here? Especially it would be of interest how `libsys.a` was generated.

There was a similar problem in #168; there it turned out that the "xc" compiler (which was built before) was not found, and this was because of a missing `"$HOME/.iraf/bin` in the `$PATH` environment variable. Not sure why this happend (usually this should be set in `$HOME/.iraf/setup.sh` which is invoked in `util/mksysgen` before the actual compilation), but pragmatically adding this may help.

**Reply by janerigby** [reply]
Thanks! I did use the tarball. I see "xc" in the path.    I tried to attach the full log below, but it's too long.  I put the log here:
https://drive.google.com/file/d/1NYiHaDW2ZwdMuRzAD_JuQ5_0IsDaml5M/view?usp=sharing

Thanks in advance!

**Reply by olebole** [reply]
I think the problem is as in #168 that there is some issue with the `$PATH`: From the log, the "xc.e" compiler is built:
```
----------------------- XC  ----------------------------
+ gcc -c -g -Wall -O2 -I/Users/jrrigby1/Applications/iraf-2.17/include -DNOVOS xc.c
+ gcc xc.o /Users/jrrigby1/Applications/iraf-2.17/unix/hlib/libboot.a /Users/jrrigby1/Applications/iraf-2.17/unix/hlib/libos.a -lm -o xc.e
+ mv -f xc.e ../../hlib
+ rm -f xc.o
```
but then it is not found; there are multiple occurrences of
```
xc -c -w access.x aread.x areadb.x await.x […] zfiott.x
xc: Command not found.
```
This leads to empty libs.
Pragmatically, you can check that that there is a symbolic link 
```
$HOME/.iraf/bin/xc --> /Users/jrrigby1/Applications/iraf-2.17/unix/bin/xc.e
```
and then just add `$HOME/.iraf/bin` to your `PATH`. In principle, this addition to `PATH` should already in https://github.com/iraf-community/iraf/blob/db7d980654160c923fe195fd9e1538bfcb60b632/unix/hlib/setup.sh#L17-L17

which is called in https://github.com/iraf-community/iraf/blob/db7d980654160c923fe195fd9e1538bfcb60b632/util/mksysgen#L7-L16

and mksysgen is actually used to build the system in the https://github.com/iraf-community/iraf/blob/db7d980654160c923fe195fd9e1538bfcb60b632/Makefile#L16-L20

but I have no idea why this doesn't work in your case. The easiest is to manually prepend  `$HOME/.iraf/bin` to your `PATH`.

**Reply by olebole** [reply]
Sorry, I only at the end have read that the command "xc" is available. Is this the correct one? Or could this be one from an earlier installation? Does `which xc` show the correct place (`$HOME/.iraf/bin/xc`), and does `xc -V` run?
```
$ xc -V
IRAFNET XC V2.4 Jan 21 2010
```

**Comment by janerigby** [comment, marked as answer]
My fabulous sysadmin David F. figured out the issue. I was dropping from my native .tcsh shell to .bash, which is a habit from installing anaconda packages. But running make sysgen in bash was messing things up, because it wasn't my native shell, and the installation has you drop in and out of the shell.

Here's what worked:
```
cp -p /Users/jrrigby1/Applications/iraf-2.17/unix/bin.macintel/xc.e /tmp/bin/xc # make a shortcut to xc
set path=( $HOME/.iraf/bin $path /tmp/bin ) # For iraf install 1/2022 # make sure xc is in the path
```
(in a new tcsh window)
```
make sysgen
```

**Reply by olebole** [reply]
Hmm, not sure if I understand this: Usually the login shell shouldn't matter. All scripts in IRAF (now) use the POSIX shell, `/bin/sh`. This is also used in `util/mksysgen`, and explicitly specified in the first line of the script.

**Reply by olebole** [reply]
Could I ask you to help me reproduce the problem? On a Mac (macOS 11.6.2), I had it building even with tcsh as a default shell:

 * called `chsh`, I set the default shell to `/bin/tcsh`,
 * opened a new terminal, which now has tcsh as shell,
 * executed `bash` to switch to the Bash,
 * invoked `./install`, hitting <Enter> on all questions,
 * did a `make macintel`
 * did a `make sysgen 2>&1 | tee build.log`

and I got a fully working IRAF. Without switching to bash (just from the tcsh), it also compiled fine. Was there any specifics in your login files?
